### ISSUE TITLE
Watch did stop on Test Cluster

### ISSUE THREAD

**Issue by jfaltermeier** [eclipsesource/theia-cloud#32]
The watch for workspace events has stopped on the test cluster. There was no exception logged. 

Our manual timer based watch logged some exceptions (see below), so it might have something to do with the connection to the cluster API. 

We should try to add more logging to figure out how to detect this state, so that we may set up the watches again. 

```
io.fabric8.kubernetes.client.KubernetesClientException: Operation: [list]  for kind: [TemplateSpecResource]  with name: [null]  in namespace: [theiacloud]  failed.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:130) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:413) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:402) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:83) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at org.eclipse.theia.cloud.operator.KillAfterRunnable.run(KillAfterRunnable.java:60) [operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.net.ConnectException: Failed to connect to /10.52.0.1:443
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:249) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.execute(RealCall.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl.send(OkHttpClientImpl.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.retryWithExponentialBackoff(OperationSupport.java:575) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:554) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:519) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:503) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:131) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	... 10 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]
	at java.net.AbstractPlainSocketImpl.doConnect(Unknown Source) ~[?:?]
	at java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source) ~[?:?]
	at java.net.AbstractPlainSocketImpl.connect(Unknown Source) ~[?:?]
	at java.net.SocksSocketImpl.connect(Unknown Source) ~[?:?]
	at java.net.Socket.connect(Unknown Source) ~[?:?]
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:129) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientBuilderImpl$InteceptorAdapter.intercept(OkHttpClientBuilderImpl.java:62) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at okhttp3.RealCall.execute(RealCall.java:93) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl.send(OkHttpClientImpl.java:138) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.retryWithExponentialBackoff(OperationSupport.java:575) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:554) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:519) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:503) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:131) ~[operator-0.7.0-SNAPSHOT-jar-with-dependencies.jar:?]
	... 10 more
```

**Comment by jfaltermeier** [comment]
Might be related: https://github.com/fabric8io/kubernetes-client/issues/3177

**Comment by jfaltermeier** [comment]
From what I saw in the logging on a local minikube, the reconnecting call might be an entry point to restart the watch.

**Comment by jfaltermeier** [comment]
Update: On the Test Cluster with 1.22 it is still working after the reconnecting calls are logged. So I believe the issue is fixed with 1.22
